src/config/initializers/apn.ts

```diff
@@ -0,0 +1,146 @@
+    const apnKeyId = process.env.APN_KEY_ID;
+    const apnTeamId = process.env.APN_TEAM_ID;
+
+    if (!apnSecret || !apnKeyId || !apnTeamId) {
+      Logger.getInstance().warn('APN env variables must be defined');
+    }
+
+    const client = http2.connect(this.apnUrl);
+
+    client.on('error', (err) => {
+      Logger.getInstance().error('APN connection error', err);
+    });
+    client.on('close', () => {
+      Logger.getInstance().warn('APN connection was closed');
+      this.handleReconnect();
+    });
+    client.on('connect', () => {
+      Logger.getInstance().info('Connected to APN');
+    });
+
+    return client;
+  }
+
+  private generateJwt(): string {
+    if (!process.env.APN_SECRET || !process.env.APN_KEY_ID || !process.env.APN_TEAM_ID) {
+      throw new Error('Undefined APN env variables, necessary for JWT generation');
+    }
+    if (this.jwt && Date.now() - this.jwtGeneratedAt < 3600) {
+      return this.jwt;
+    }
+
+    this.jwt = jwt.sign(
+      {
+        iss: process.env.APN_TEAM_ID,
+        iat: Math.floor(Date.now() / 1000),
+      },
+      Buffer.from(process.env.APN_SECRET as string, 'base64').toString('utf8'),
+      {
+        algorithm: 'ES256',
+        header: {
+          alg: 'ES256',
+          kid: process.env.APN_KEY_ID,
+        } as JwtHeader,
+      },
+    );
+
+    this.jwtGeneratedAt = Date.now();
+
+    return this.jwt;
+  }
+
+  private handleReconnect() {
+    if (this.reconnectAttempts < this.maxReconnectAttempts) {
+      setTimeout(() => {
+        Logger.getInstance().info(`Attempting to reconnect to APN (#${this.reconnectAttempts + 1})`);
+        this.connectToAPN();
+        this.reconnectAttempts++;
+      }, this.reconnectDelay * Math.pow(2, this.reconnectAttempts));
+    } else {
+      Logger.getInstance().error('Maximum APN reconnection attempts reached');
+    }
+  }
+
+  public sendStorageNotification(deviceToken: string, userUuid: string): Promise<{ statusCode: number; body: string }> {
+    return new Promise((resolve, reject) => {
+      if (!this.client || this.client.closed) {
+        this.connectToAPN();
+      }
+      const headers = {
+        'apns-topic': `${this.bundleId}.pushkit.fileprovider`,
+        authorization: `bearer ${this.generateJwt()}`,
+      };
+
+      const options = {
+        ':method': 'POST',
+        ':path': `/3/device/${deviceToken}`,
+        ':scheme': 'https',
+        ':authority': 'api.push.apple.com',
+        'content-type': 'application/json',
+      };
+
+      const req = this.client.request({ ...options, ...headers });
+
+      req.setEncoding('utf8');
+      req.write(
+        JSON.stringify({
+          'container-identifier': 'NSFileProviderWorkingSetContainerItemIdentifier',
+          domain: userUuid,
+        }),
+      );
+
+      req.end();
+
+      let statusCode = 0;
+      let data = '';
+
+      req.on('response', (_, status) => {
+        statusCode = status || 0;
+      });
+
+      req.on('data', (chunk) => {
```

Comment:
This event-based request usage is not the best way to handle this consistently, why not use `axios` if you are not streaming anything?

Reply:
Axios doesn't support http/2, and APN only accpets connections over this protocol